### PR TITLE
Queries Panel: run local query from panel

### PR DESCRIPTION
**Pull request by norascheuch**
Adds a 'run'-icon when hovering over a file in the queries panel when a local database is selected. Runs the query when clicked.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/queries-panel/queries-panel.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-panel.ts]: not relevant anymore

File: extensions/ql-vscode/src/queries-panel/queries-panel.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/queries-panel/queries-panel.ts:35]: It would be safer and more correct to use `Uri.file` here:
```suggestion
      vscode.Uri.file(queryTreeViewItem.path),
```

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/queries-panel/queries-module.ts:30]: If we're going to use this check in multiple places, can we pull it out to a function so we don't duplicate the logic or the comment. Maybe into a `shouldInitializeQueriesPanel` method.

Sadly the comment is slightly outdated already because of the merge of https://github.com/github/vscode-codeql/pull/2528, so that's a perfect example of why it's good to avoid duplication.

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/queries-panel/queries-module.ts:33]: I don't particularly love that we have to do this cast. We shouldn't have to force the type system, so I think it shows that things perhaps aren't quite right at a high-level.

Earlier today we were discussing always registering the command and then only showing the panel if the feature flag is enabled. That would mean something like https://github.com/github/vscode-codeql/commit/42886c69aed49878d66a4e298e0a6829de00cee7, though perhaps I didn't make it clear enough what I meant. What do you think? If you don't think that's better then that's ok, but I'd be interested to hear the reasoning.

Or as another alternative would it be better providing dummy implementations that just log an error? That way if one of these commands does accidentally become callable then the user will get a nicer error message.

File: extensions/ql-vscode/src/common/app.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/common/app.ts:21]: Should get a second opinion about exposing the `queryServerCommands` field on the app interface. @koesie10, do you have any thoughts? Is there a reason not to, or did we just not need it until this point?

**Review by robertbrignull — COMMENTED** [review]
I think it's along the right lines, though I have a couple of small comments and I'd like to get another opinion on things since I suggested some parts.

File: extensions/ql-vscode/src/queries-panel/queries-panel.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/queries-panel/queries-panel.ts:35]: Good point! The "Run query" button currently fails for me (on Windows) with:

image


Just tried @robertbrignull's suggestion though, and can confirm that it works 🎉

**Review by shati-patel — COMMENTED** [review]
Looks good overall 😍 Tested locally and left some comments too 📝

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-module.ts:33]: Ah, yes. My thought was to initialize as little as possible. I agree that the cast is a code smell, but to push it all to the panel and initialize that didn't sit well with me either.

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-module.ts]: 👍 I extracted the check into a method. We can discuss where to push the guard in your other comment.

File: extensions/ql-vscode/src/queries-panel/queries-panel.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-panel.ts]: 👍

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-module.ts]: 👍 extracted this out to a method

File: extensions/ql-vscode/src/common/app.ts
Comment (koesie10) [reply extensions/ql-vscode/src/common/app.ts:21]: I think we just haven't needed it until this point. I don't see a particular reason not to expose it, it's coming from the same file so we aren't importing anything extra.

File: extensions/ql-vscode/src/common/commands.ts
Comment (koesie10) [inline extensions/ql-vscode/src/common/commands.ts:15]: This should be a `type` import to ensure this file remains somewhat independent of `vscode` and only needs the types from `vscode`, rather than the complete `vscode` module:

```suggestion
import type { QueryTreeViewItem } from "../queries-panel/query-tree-view-item";
```

File: extensions/ql-vscode/src/common/commands.ts
Comment (koesie10) [inline extensions/ql-vscode/src/common/commands.ts:133]: Is there a reason for the separation of these two commands? They seem pretty related to me and I would expect both `codeQLQueries` commands to be in the same type.

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (koesie10) [reply extensions/ql-vscode/src/queries-panel/queries-module.ts:31]: We have some other optional commands as well, the `TestUICommands` commands. These are declared as `Partial<TestUICommands>` in the `commands.ts` file so that they aren't required to be set. I think that would work well here as well.

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (koesie10) [inline extensions/ql-vscode/src/queries-panel/queries-module.ts:44]: This is not calling the method, but is checking that the method exists. It should be (also in the other instances):

```suggestion
    if (!this.shouldInitializeQueriesPanel()) {
```

You can also change the `shouldInitializeQueriesPanel` to be a getter, which would be my preferred solution in this case since there are no real side-effects.

File: extensions/ql-vscode/src/queries-panel/queries-panel.ts
Comment (koesie10) [inline extensions/ql-vscode/src/queries-panel/queries-panel.ts:35]: My preference would be to import `Uri` from `vscode` directly, so that we could call this instead:

```suggestion
     Uri.file(queryTreeViewItem.path),
```

This is how we're doing this in most other files as well.

File: extensions/ql-vscode/package.json
Comment (shati-patel) [inline extensions/ql-vscode/package.json:1110]: Maybe a silly question: do we need `codeQL.currentDatabaseItem`?

If you don't have a current database, the "Run Local Query" command pops up a window and asks you to select one anyway. (I don't think it matters either way, just wondering 🙃)

Comment (norascheuch) [reply extensions/ql-vscode/package.json:1110]: Hm, no reason given. I just thought this was the requirement.

File: extensions/ql-vscode/src/queries-panel/queries-module.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-module.ts:44]: Thanks for catching this.

File: extensions/ql-vscode/src/queries-panel/queries-panel.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/queries-panel/queries-panel.ts:35]: 👍 I used vscode here bc line 18, but adjusted both now

File: extensions/ql-vscode/src/common/commands.ts
Comment (norascheuch) [reply extensions/ql-vscode/src/common/commands.ts:133]: The command here is a new replica of the `runQuery` command that is defined in the local-queries and therefore defined here in `LocalQueryCommands`. It is executed in the queries-panel when a user 'executes' `"codeQLQueries.runLocalQueryContextInline"` through the panel. That's why we have 2 different commands. 

I suppose we could adjust the original implementation in `runQuery` to handle the case when no uri is supplied but a QueryPanelTreeViewItem instead and leave the queries-panel out of it?..

File: extensions/ql-vscode/package.json
Comment (shati-patel) [reply extensions/ql-vscode/package.json:1106]: We can probably remove it in that case, to slightly simplify the `when` 🤷🏽

Comment (norascheuch) [reply extensions/ql-vscode/package.json:1106]: After discussing this and checking the flow for both cases we decided to keep the db check.

File: extensions/ql-vscode/package.json
Comment (shati-patel) [inline extensions/ql-vscode/package.json:505]: We still want this command to be prefixed with `codeQLQueries` instead of `codeQL`, since its a command that is scoped to a specific CodeQL-controlled panel (instead of being generally available in the file editor/command palette). 

(This would match the other commands in `"view/item/context"` too 😎 )

File: extensions/ql-vscode/test/unit-tests/command-lint.test.ts
Comment (shati-patel) [inline extensions/ql-vscode/test/unit-tests/command-lint.test.ts:73]: As mentioned above, I think the previous change was correct (deleting this bit, but adding `command.match(/^codeQLQueries\./)` to the "allowed prefixes"). 

We currently add a new prefix for each new panel, so this would be consistent with the others 😺

**Review by shati-patel — COMMENTED** [review]
> It seems the command lint test was not quite up to date, I adjusted a bit of it to pass the tests sensibly for the current change.

I've left some comments about the command naming, but otherwise this PR looks nice and tidy 🙌🏽

File: extensions/ql-vscode/package.json
Comment (norascheuch) [reply extensions/ql-vscode/package.json:505]: Ah, yes 🤦‍♀️

**Review by robertbrignull — APPROVED** [review]
This looks ok to me, though I wonder if in the long run this is introducing coupling that we don't want. I'm happy for it to be merged, depending on what @koesie10 thinks.

Putting the `CodeQLQueries.runLocalQueryFromQueriesPanel` command on the `LocalQueries` class certainly is the least complex solution right now. It avoids some problem like the queries panel not being initialised, though it may be we'll run into that again for the next command we want to add and it'll be harder to solve that time. Long term I wonder if `LocalQueries` should depend on the queries panel, or if it should be more the other way around. In other component we pass in `LocalQueries` as an argument.

I'd also be happy to go for (and maybe have a slight preference for) the [suggestion](https://github.com/github/vscode-codeql/pull/2534#discussion_r1239617305) of using `Partial` to get around the initialisation problem, and thus keeping the command definition in `QueriesPanel`. To make this work we could expose the `LocalQueries.runQuery` method publicly so it can call that. All that method does is wrap the public `compileAndRunQuery` method with a progress bar, so making `runQuery` public too seems ok.

**Comment by robertbrignull** [comment]
I think let's merge this so we unblock this work, but keep in mind the idea of moving the commands to `QueriesPanel` so it makes the dependency go in the other direction which is probably a little cleaner. We can make that future change at any point, so opening a tech-debt issue would be fine.

**Comment by norascheuch** [comment]
Ok, will open the tech dept issue @robertbrignull . I'm wondering how we could make this process more efficient though. All three options went through my head while I was working on this at some point. How do I know when is a good way to discuss these things before the work is finished and would need to be redone?

**Pull request merged**
